README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ggblur <img src="man/figures/logo-blurred.png" align="right" height=230/>

<!-- badges: start -->

![](http://img.shields.io/badge/cool-useless-green.svg)
<!-- badges: end -->

`ggblur` provides `geom_point_blur()` for use in `ggplot` - this geom
allows you to control the blurriness of the drawn points, but is
otherwise identical to the standard `geom_point()`.

#### What’s in the box?

  - `geom_point_blur()`
      - same as `geom_point()` but now accepts `blur_size` as an
        aesthetic
      - also allows for control of the smoothness of the blur
        (`blur_steps`)
  - `scale_blur_size_continuous()`, `scale_blur_size_discrete()` and
    `scale_blur_size_manual()` for controlling `blur_size` when used as
    a mapped aesthetic/

#### Similar packages

  - [ggecho](https://github.com/coolbutuseless/ggecho) is an earlier
    experiment of mine from 2019 where I echo components using a custom
    stat `stat_echo`. This solution works across multiple geoms, but
    didn’t allow for adjusting the echo for individual points.
  - [ggfx](https://github.com/thomasp85/ggfx) adds blur at the layer
    level i.e.  no individual control over the amount of blur for each
    point.

#### How it works

Blur is simulated by keeping the original point, and drawing a sequence
of larger, faded points behind it (illustrated below).

The number of points rendered is controlled by `blur_steps` and the
distance to which the blur extends beyond the edge of the original point
is controlled by `blur_size`.

Note: Paul Murrell has some experiments to add radial gradients to grid
graphics (see [his github](https://github.com/pmur002/r-defs-proposal))
and if this becomes part of grid it may be a better method of simulating
blur.

<div>

<img src="man/figures/point.png" width="45%" align="left">
<img src="man/figures/how.png" width="45%" align="left">

</div>

<div style="clear: both;" />

## Installation

You can install the development version from
[GitHub](https://github.com/coolbutuseless/ggblur) with:

``` r
# install.packages("devtools")
devtools::install_github("coolbutuseless/ggblur")
```

## Example 1 - constant `blur_size`

When not used as a mapped aesthetic within `aes()`, the same `blur_size`
is applied to every point.

``` r
library(ggplot2)
library(ggblur)

ggplot(mtcars) +
  geom_point_blur(aes(mpg, wt), blur_size = 10) +
  theme_bw() + 
  labs(title = "Same blur for each point")
```

<img src="man/figures/README-example1-1.png" width="100%" />

## Example 2 - mapping `blur_size` as an aesthetic

When used as a mapped aesthetic within `aes()`, the `blur_size` is
calculated individually for each point.

``` r
library(ggplot2)
library(ggblur)

ggplot(mtcars) +
  geom_point_blur(aes(mpg, wt, blur_size = disp)) +
  theme_bw() + 
  labs(title = "Larger blur indicates larger engine displacement")
```

<img src="man/figures/README-example2-1.png" width="100%" />

## Example 3 - control over blur parameters

`blur_steps` and `scale_blue_size_continuous/discrete/manual()` can be
used to further customise the appearance.

``` r
ggplot(mtcars) +
  geom_point_blur(aes(mpg, wt, blur_size = disp), blur_steps = 3) +
  scale_blur_size_continuous(range = c(1, 15)) +
  theme_bw() + 
  labs(title = "Larger blur indicates larger engine displacement")
```

<img src="man/figures/README-example3-1.png" width="100%" />

## Example 4 - blur with colour

``` r
ggplot(mtcars) +
  geom_point_blur(aes(mpg, wt, blur_size = disp, colour = as.factor(cyl))) +
  scale_blur_size_continuous(range = c(1, 15)) +
  theme_bw() + 
  labs(title = "Larger blur indicates larger engine displacement")
```

<img src="man/figures/README-example4-1.png" width="100%" />


NEWS.md

# ggblur 0.1.2  2020-02-15

* Reworked how a sequence of individual overlapping alphas are created in order
  to achieve a more consistent blur appearance.
* Removed `blur_alpha` as a parameter. Alpha is now derived from that of the 
  parent point object. This does not currently support a different alpha per-point,
  but assumes that alpha is constant across all points.

# ggblur 0.1.1  2020-02-11

* Added legend support

# ggblur 0.1.0  2020-02-09

* Initial release


R/alpha.R


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Calculate the cumulative alpha as the individual alpha are combined
#'
#' Remember:
#'
#' \itemize{
#' \item{alpha is not additive}
#' \item{transmissivity = 1 - alpha}
#' \item{transmissivity is multiplicative}
#' }
#'
#' @param individual_alpha numeric vector of alpha values which will be
#'        superimposed.
#'
#' @return numeric vector of the cumulative alpha as these elements are
#'         rendered on top of each other
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
calc_cumulative_alpha <- function(individual_alpha) {
  1 - cumprod(1 - individual_alpha)
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Calculate the individual alpha levels to recreate the cumulative alpha series
#'
#'
#' Remember:
#'
#' \itemize{
#' \item{alpha is not additive}
#' \item{transmissivity = 1 - alpha}
#' \item{transmissivity is multiplicative}
#' }
#'
#'
#' @param cumulative_alpha numeric vector of desired cumulative alpha as
#'        multiple objects are rendered on top of each other
#'
#' @return numeric vector of the individual alpha values of the objects to be
#'         superimposed
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
calc_individual_alpha <- function(cumulative_alpha) {
  N <- length(cumulative_alpha)
  individual_alpha    <- numeric(N)
  individual_alpha[1] <- cumulative_alpha[1]
  for (i in seq(N-1)) {
    individual_alpha[i+1] <- 1 - (1 - cumulative_alpha[i+1])/(1 - cumulative_alpha[i])
  }
  individual_alpha
}


R/ggplot-compat.R


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# For compatibility with ggplot2 - and avoiding dplyr/magrittr import
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
'%||%' <- function(x, y) {
  if (is.null(x)) {
    y
  } else {
    x
  }
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# For compatibility with ggplot2
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
new_data_frame <- function(x) {data.frame(x, stringsAsFactors = FALSE)}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# For compatibility with ggplot2
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ggname <- function(prefix, grob) {
  grob$name <- grid::grobName(grob, prefix)
  grob
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# For compatibility with ggplot2
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
manual_scale <- function(aesthetic, values = NULL, breaks = waiver(), ...) {
  # check for missing `values` parameter, in lieu of providing
  # a default to all the different scale_*_manual() functions
  force(values)

  # order values according to breaks
  if (is.vector(values) && is.null(names(values)) && !inherits(breaks, "waiver") &&
      !is.null(breaks)) {
    if (length(breaks) != length(values)) {
      stop(paste("
        Differing number of values and breaks in manual scale.",
        length(values), "values provided compared to",
        length(breaks), "breaks.
      "))
    }
    names(values) <- breaks
  }

  pal <- function(n) {
    if (n > length(values)) {
      stop(paste("Insufficient values in manual scale.", n,
                 "needed but only",
                 length(values),
                 "provided."))
    }
    values
  }
  ggplot2::discrete_scale(aesthetic, "manual", pal, breaks = breaks, ...)
}


R/scale-blur-size.R


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Scale blur size
#'
#' @param guide default: 'legend'.
#' @param name,breaks,labels,limits,range,trans,values,... see \code{?ggplot2::scale_size()}
#'
#' @rdname scale_blur_size
#'
#' @import ggplot2
#' @import scales
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
scale_blur_size_continuous <- function(name = waiver(), breaks = waiver(), labels = waiver(),
                         limits = NULL, range = c(3, 9),
                         trans = "identity", guide = 'legend') {
  ggplot2::continuous_scale(
    aesthetics = "blur_size",
    scale_name = "blur_size",
    palette    = scales::rescale_pal(range), name = name,
    breaks = breaks, labels = labels, limits = limits, trans = trans,
    guide = guide)
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' @rdname scale_blur_size
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
scale_blur_size <- scale_blur_size_continuous


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' @rdname scale_blur_size
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
scale_blur_size_discrete <- function(..., range = c(4, 9), guide='legend') {
  force(range)

  ggplot2::discrete_scale(
    aesthetics = "blur_size",
    scale_name = "blur_size_d",
    palette    = function(n) {
      seq(range[1], range[2], length.out = n)
    },
    guide = guide,
    ...
  )
}



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' @rdname scale_blur_size
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
scale_blur_size_manual <- function(..., values) {
  manual_scale('blur_size', values, ...)
}







R/draw-key-point-blur.R


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Key glyphs for legends
#'
#' Each geom has an associated function that draws the key when the geom needs
#' to be displayed in a legend. These functions are called `draw_key_*()`, where
#' `*` stands for the name of the respective key glyph. The key glyphs can be
#' customized for individual geoms by providing a geom with the `key_glyph`
#' argument (see [`layer()`] or examples below.)
#'
#' @param data A single row data frame containing the scaled aesthetics to
#'   display in this key
#' @param params A list of additional parameters supplied to the geom.
#' @param size Width and height of key in mm.
#'
#' @return A grid grob.
#'
#' @import grid
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
draw_key_point_blur <- function(data, params, size) {
  if (is.null(data$shape)) {
    data$shape <- 19
  } else if (is.character(data$shape)) {
    data$shape <- translate_shape_string(data$shape)
  }

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Just rename the 'data' variable to 'coords' so the code is then
  # pretty much identical to that in the 'geom'
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  coords <- data

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # How many steps in the blur process?
  # What should be the alpha of an individual step?
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  blur_steps <- coords$blur_steps[1]

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Create a blur grob somewhere between [0,1] * blur_size.
  # Ensure lwd = 0 so that no outer stroke is included.
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  create_blur_grob <- function(step) {

    fraction   <- fractions[step]
    this_alpha <- ind_alpha[step]

    grid::pointsGrob(
      x = 0.5, y = 0.5,
      pch = coords$shape,
      gp = grid::gpar(
        col      = alpha(coords$colour, this_alpha),
        fill     = alpha(coords$fill  , this_alpha),
        fontsize = (coords$size + coords$blur_size * fraction) * .pt +
          coords$stroke * .stroke / 2,
        lwd      = 0
      )
    )
  }

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Gaussian-ish sequence of blurring steps
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  fractions <- qnorm(seq(0.95, 0.5, length.out = blur_steps + 1))
  fractions <- fractions/max(fractions)
  fractions <- head(fractions, -1)

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Sequence of individual alphas
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  top_alpha <- coords$alpha[1]
  if (is.null(top_alpha) || is.na(top_alpha)) {
    top_alpha <- 1
  }
  cumulative_alpha <- seq(0.1, top_alpha, length.out = blur_steps + 1)
  cumulative_alpha <- head(cumulative_alpha, -1)
  ind_alpha  <- calc_individual_alpha(cumulative_alpha)

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Create a sequence of blur steps as grobs and package as a grobTree
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  blur_grobs <- lapply(seq_along(fractions), create_blur_grob)
  blur_grobs <- do.call(grid::grobTree, blur_grobs)


  grid::grobTree(
    blur_grobs,
    grid::pointsGrob(
      x = 0.5, y = 0.5,
      pch = data$shape,
      gp = grid::gpar(
        col      = alpha(data$colour %||% "black", data$alpha),
        fill     = alpha(data$fill   %||% "black", data$alpha),
        fontsize = (data$size %||% 1.5) * .pt + (data$stroke %||% 0.5) * .stroke / 2,
        lwd      = (data$stroke %||% 0.5) * .stroke / 2
      )
    )
  )
}


R/geom-point-blur.R

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Draw blurry points in ggplot
#'
#' @param blur_size How far should the blur extend from the edge of the drawn point.
#'        Default: 3
#' @param blur_steps Number of repetitions to create blur. A higher value for
#'        \code{blur_steps} will results in a smoother looking blur. Default: 20
#' @param mapping,data,stat,position,...,na.rm,show.legend,inherit.aes See
#' documentation for \code{ggplot2::geom_point()}
#'
#' @import ggplot2
#' @export
#'
#' @examples
#' \dontrun{
#' ggplot(mtcars) +
#' geom_point_blur(aes(mpg, wt, blur_size = disp), blur_steps = 2) +
#'   scale_blur_size_continuous(range = c(1, 15)) +
#'   theme_bw() +
#'   labs(title = "Larger blur indicates larger engine displacement")
#' }
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
geom_point_blur <- function(mapping = NULL, data = NULL,
                            stat = "identity", position = "identity",
                            ...,
                            blur_size   = 3,
                            blur_steps  = 20,
                            na.rm       = FALSE,
                            show.legend = NA,
                            inherit.aes = TRUE) {
  ggplot2::layer(
    data = data,
    mapping     = mapping,
    stat        = stat,
    geom        = GeomPointBlur,
    position    = position,
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params      = list(
      na.rm      = na.rm,
      blur_steps = blur_steps,
      ...
    )
  )
}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' GeomPointBlur
#'
#' @rdname ggplot2-ggproto
#' @format NULL
#' @usage NULL
#'
#' @importFrom stats qnorm
#' @importFrom utils head
#' @import grid
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
GeomPointBlur <- ggproto(
  "GeomPointBlur", Geom,
  required_aes = c("x", "y"),
  non_missing_aes = c("size", "shape", "colour"),
  default_aes = ggplot2::aes(
    shape      = 19,
    colour     = "black",
    size       = 1.5,
    fill       = NA,
    alpha      = NA,
    stroke     = 0.5,
    blur_size  = 3,
    blur_steps = 20
  ),

  draw_panel = function(data, panel_params, coord, na.rm = FALSE) {
    if (is.character(data$shape)) {
      data$shape <- translate_shape_string(data$shape)
    }
    coords <- coord$transform(data, panel_params)

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # How many steps in the blur process?
    # What should be the alpha of an individual step?
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    blur_steps <- coords$blur_steps[1]

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Create a blur grob somewhere between [0,1] * blur_size.
    # Ensure lwd = 0 so that no outer stroke is included.
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    create_blur_grob <- function(step) {

      fraction   <- fractions[step]
      this_alpha <- ind_alpha[step]

      grid::pointsGrob(
        coords$x, coords$y,
        pch = coords$shape,
        gp = grid::gpar(
          col      = alpha(coords$colour, this_alpha),
          fill     = alpha(coords$fill  , this_alpha),
          fontsize = (coords$size + coords$blur_size * fraction) * .pt +
            coords$stroke * .stroke / 2,
          lwd      = 0
        )
      )
    }

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Gaussian-ish sequence of blurring steps
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    fractions <- qnorm(seq(0.95, 0.5, length.out = blur_steps + 1))
    fractions <- fractions/max(fractions)
    fractions <- head(fractions, -1)

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Sequence of individual alphas
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    top_alpha <- coords$alpha[1]
    if (is.null(top_alpha) || is.na(top_alpha)) {
      top_alpha <- 1
    }
    cumulative_alpha <- seq(0.1, top_alpha, length.out = blur_steps + 1)
    cumulative_alpha <- head(cumulative_alpha, -1)
    ind_alpha  <- calc_individual_alpha(cumulative_alpha)

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Create a sequence of blur steps as grobs and package as a grobTree
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    blur_grobs <- lapply(seq_along(fractions), create_blur_grob)
    blur_grobs <- do.call(grid::grobTree, blur_grobs)

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # the returned grob is now a grob tree with the blur steps behind the
    # final drawn points.
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ggname(
      "geom_point",
      grid::grobTree(
        blur_grobs,
        grid::pointsGrob(
          coords$x, coords$y,
          pch = coords$shape,
          gp = grid::gpar(
            # Stroke is added around the outside of the point
            col      = alpha(coords$colour, coords$alpha),
            fill     = alpha(coords$fill, coords$alpha),
            fontsize = coords$size * .pt + coords$stroke * .stroke / 2,
            lwd      = coords$stroke * .stroke / 2
          )
        )
      )
    )
  },

  draw_key = draw_key_point_blur
)




if (FALSE) {

  library(ggplot2)

  (p <- ggplot(mtcars) +
      geom_point(aes(mpg, wt, blur_size = disp), blur_steps = 20) +
      # scale_blur_size_continuous(range = c(0, 10)) +
      theme_bw()
  )

  pdf("working/test.pdf")
  p
  dev.off()

}









